src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs

```diff
@@ -1846,55 +1848,121 @@ private static void CheckStringSplitOptions(StringSplitOptions options)
+            // extension from 32- to 64-bit is zero-extending rather than sign-extending. The end result
+            // of this is that if either input is negative or if the input sum overflows past Int32.MaxValue,
+            // that information is captured correctly in the comparison against the backing _length field.
+            // We don't use this same mechanism in a 32-bit process due to the overhead of 64-bit arithmetic.
+            if ((ulong)(uint)startIndex + (ulong)(uint)length > (ulong)(uint)thisLength)
+                ThrowHelper.ThrowArgumentOutOfRangeException();
+#else
+            if ((uint)startIndex > (uint)thisLength || (uint)length > (uint)(thisLength - startIndex))
+                ThrowSubstringArgumentOutRangeException(startIndex, length);
+#endif
 
             if (length == 0)
             {
                 return string.Empty;
             }
 
-            if (startIndex == 0 && length == this.Length)
+            if (length == thisLength)
             {
                 return this;
             }
 
             return InternalSubString(startIndex, length);
         }
 
+        [DoesNotReturn]
+        private void ThrowSubstringArgumentOutRangeException(int startIndex, int length)
+        {
+            if (startIndex < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(startIndex), SR.ArgumentOutOfRange_StartIndex);
+            }
+            if (startIndex > Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(startIndex), SR.ArgumentOutOfRange_StartIndexLargerThanLength);
+            }
+            if (length < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(length), SR.ArgumentOutOfRange_NegativeLength);
+            }
+            // Check not needed since we only call this when startIndex, length are out of range
+            //if (startIndex > Length - length)
+            throw new ArgumentOutOfRangeException(nameof(length), SR.ArgumentOutOfRange_IndexLength);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         private string InternalSubString(int startIndex, int length)
         {
             Debug.Assert(startIndex >= 0 && startIndex <= this.Length, "StartIndex is out of range!");
             Debug.Assert(length >= 0 && startIndex <= this.Length - length, "length is out of range!");
 
             string result = FastAllocateString(length);
 
-            Buffer.Memmove(
-                elementCount: (uint)result.Length, // derefing Length now allows JIT to prove 'result' not null below
-                destination: ref result._firstChar,
-                source: ref Unsafe.Add(ref _firstChar, (nint)(uint)startIndex /* force zero-extension */));
-
+            ref char source = ref Unsafe.Add(ref _firstChar, (nint)(uint)startIndex); /* force zero-extension */
+            ref char destination = ref result._firstChar;
+            if (length < 16)
+            {
```

Comment:
I expect that this will regress performance on some types of current or future hardware, in some cases at least.

I do not think it is useful to try to beat the MemMove implementation by partially inlining it.

Reply:
@jkotas thanks for the comment. Yeah, I thought that might be the case, it does show an improvement of ~10% for small strings, but that may not be enough to justify the added code. In that case only the guard clause change makes sense, and that only has minor impact. Although improves jit'ed code size significantly since throws are separated out.